onecodex/models/base_sample_collection.py

```diff
@@ -0,0 +1,1375 @@
+    """
+
+    def __hash__(self):
+        """To enable pickling (for lru_cache)."""
+        return hash(tuple(r.id for r in self._res_list))
+
+    def __init__(self, *args, **kwargs):
+        """Instantiate a new SampleCollection containing `Samples` or `Classifications` objects.
+
+        Parameters
+        ----------
+        objects : list
+            A list of `onecodex.models.Samples` or `onecodex.models.Classifications` objects
+            which will be processed into a `SampleCollection`
+
+        skip_missing : bool, optional
+            If an analysis was not successful, exclude it, warn, and keep going
+
+        Examples
+        --------
+        Given a list of Samples, create a new SampleCollection using abundances:
+
+            samples = [sample1, sample2, sample3]
+            collection = SampleCollection(samples, metric='abundance')
+
+        Notes
+        -----
+        To provide access to the list-like API of `ResourceList`, must also accept a list of
+        unwrapped potion resources and a One Codex model.
+        """
+
+        job = kwargs.get("job", None)
+        skip_missing = kwargs.get("skip_missing", True)
+        objects = [*args[0]]
+
+        if not all(
+            [isinstance(obj, Samples) or isinstance(obj, Classifications) for obj in objects]
+        ):
+            raise OneCodexException(
+                "SampleCollection can only contain One Codex Samples or Classifications objects"
+            )
+
+        # are they all the same model?
+        object_classes = [type(obj) for obj in objects]
+
+        if len(set(object_classes)) > 1:
+            raise OneCodexException(
+                "SampleCollection can contain Samples or Classifications, but not both"
+            )
+
+        model = objects[0].__class__ if len(objects) > 0 else Samples
+
+        self._kwargs = {
+            "skip_missing": skip_missing,
+            "job": job,
+        }
+        self._oc_model = model
+        self._res_list = objects
+
+    def _check_valid_resource(self, other, check_for_dupes=True):
+        if not isinstance(other, list):
+            other = [other]
+
+        other_ids = []
+        for o in other:
+            if not isinstance(o, self._oc_model):
+                raise ValueError(
+                    "Expected object of type '{}', got '{}'".format(
+                        self._oc_model.__name__, type(o).__name__
+                    )
+                )
+
+            other_ids.append(o.id)
+
+        if check_for_dupes:
+            # duplicates are not allowed
+            self_ids = [s.id for s in self._res_list]
+
+            if len(set(self_ids + other_ids)) != len(self_ids + other_ids):
+                raise OneCodexException(
+                    "{} cannot contain duplicate objects".format(self.__class__.__name__)
+                )
+
+    def __eq__(self, other):
+        return all(id(x) == id(y) for x, y in zip(self, other))
```

Comment:
Why `id(x) == id(y)` VS `x.id == y.id`? `id()` is basically memory address. What if both objects are the same (with same ids) but are in the different memory locations?